collections.go

```diff
@@ -23,6 +33,10 @@ type KeyEncoder[T any] interface {
 	Decode(b []byte) (int, T)
 	// Stringify returns a string representation of T.
 	Stringify(key T) string
+	// Type describes the type of key encoder.
+	Type() string
```

Comment:
Some implementations of this method return an arbitrary hard-coded string e.g. `sdk/string` or `sdk/uint64` or `sdk/AccAddress`. The implementation that returns e.g. `sdk/AccAddress` uses the AccAddress type from whichever version of github.com/cosmos/cosmos-sdk/types is imported and included in the binary, which is not well defined. That means that a value of Type() sdk/AccAddress does not necessarily have the same encoded form. Is that safe?

Reply:
It's not safe. I don't think we should have AccAddress encoders at all tbh because they depend on a global variable but that's a separate matter. Also the way types are expressed in this PR is very much a rough draft. We should come up with something better and more formalized.

Reply:
@aaronc only stringification of Acc/Val/ConsAddress depends on the global, stringification is useful only in case you want to retain humanised address ordering.

I think having a key enc which only does `Encode(addr sdk.AccAdreess) ([]byte, error) { return addr }` should be enough.

Also a length prefixed acc address encoder.

Reply:
@peterbourgon, I think differentiation of key encoder types lays within the type being encoded and the method of encoding. EG: uint64-big-endian, sdk.AccAddress-length-prefixed, etc.

Reply:
I think there is some confusion here. An interface expresses a consumer contract, that is, it does not say "I can do this", rather it says "I model an expectation of something  that can do this". So an interface called Encoder expresses a set of capabilities which types may implement. It's a bit weird to have that interface type generic across concrete types that are part of its method's signatures. If there is some type e.g. AccAddress which needs to have N different encodings (and decodings), the way to model that is to attach distinct methods to the AccAddress type for each of those encodings (and decodings) and define _independent_ methods for each of those things.

```go
func (a *AccAddress) EncodeASN1() ([]byte, error) { ... }
func (a *AccAddress) DecodeASN1(b []byte) error { ... }

type ASN1Encoder interface { EncodeASN1() ([]byte, error) }
type ASN1Decoder interface { DecodeASN1([]byte) (error) }
```

That is, an Encoder is typically a behavior that types can satisfy, it's generally not a type itself that takes values of other things that it can encode.

Reply:
@testinginprod we're going to get rid of the global eventually which means AccAddressEncoder will need to take the bech32 prefix as a parameter when that happens.

Reply:
Delegating the encoding to the encoder types is fine IMHO. I don't see a problem with this design. It does make schema reflection a little harder because we need a way of naming and encodings in an extensible - which may mean we need something like encoding registry, but that's solvable.